File: js/816.js
var isValid = function (S, i, j) {
  const s = S.slice(i, j + 1)
  if (s.length > 1 && Number(s) === 0) {
    return false
  }
  return true
}

var isValid2 = (e) => Number(e).toString().length === e.length

var gen = function (S, i, j) {
  const s = S.slice(i, j + 1)
  const result = []
  if (isValid2(s)) {
    result.push(s)
  }
  for (let k = 1; k < s.length; k++) {
    // console.log(k, s.slice(i, k), s.slice(k));
    const e = s.slice(0, k) + '.' + s.slice(k)
    if (isValid2(e)) {
      result.push(e)
    }
  }
  // console.log({ s, result });
  return result
}


/**
 * @param {string} S
 * @return {string[]}
 */
var ambiguousCoordinates = function (S) {

  S = S.slice(1, S.length - 1)
  const result = []
  for (let i = 0; i < S.length - 1; i++) {
    if (isValid(S, 0, i) && isValid(S, i + 1, S.length - 1)) {
      const left = gen(S, 0, i)
      const right = gen(S, i + 1, S.length - 1)
      for (let j = 0; j < left.length; j++) {
        for (let k = 0; k < right.length; k++) {
          result.push('(' + left[j] + ', ' + right[k] + ')')
        }
      }
    }
  }
  return result
};

var assert = require('assert');
assert.deepEqual(gen('12', 0, 1), ['12', '1.2']);
assert.deepEqual(gen('100', 0, 2), ['100']);
assert.deepEqual(ambiguousCoordinates("(123)").sort(), ["(1, 23)", "(12, 3)", "(1.2, 3)", "(1, 2.3)"].sort())
assert.deepEqual(ambiguousCoordinates("(00011)").sort(), ["(0.001, 1)", "(0, 0.011)"].sort())
assert.deepEqual(ambiguousCoordinates("(0123)").sort(), ["(0, 123)", "(0, 12.3)", "(0, 1.23)", "(0.1, 23)", "(0.1, 2.3)", "(0.12, 3)"].sort())
assert.deepEqual(ambiguousCoordinates("(100)").sort(), ["(10, 0)"].sort())

File: py/816.py

import unittest


def trim(s, c=" "):
    i = 0
    for k in range(0, len(s)):
        i = k
        if s[k] != c:
            if s[k] == ".":
                i -= 1
            break
    j = len(s)
    for k in range(len(s), 0, -1):
        j = k
        if s[k-1] != c:
            break
    return s[i:j]


class Solution:
    def isValid(self, S, i, j):
        s = S[i:j+1]
        if len(s) > 1 and float(s) == 0:
            return False
        return True

    def isValid2(self, e):
        e2 = float(e)
        if e2 == int(e2):
            return len(str(int(e2))) == len(e)
        else:

            return len(trim(e, "0")) == len(e)

    def gen(self, S, i, j):
        s = S[i:j+1]
        r = []
        if self.isValid2(s):
            r.append(s)
        for k in range(1, len(s)):
            e = s[0:k] + "." + s[k:]
            # print(e, self.isValid2(e))
            if self.isValid2(e):
                r.append(e)
        return r

    def ambiguousCoordinates(self, S):
        """
        :type S: str
        :rtype: List[str]
        """
        S = S[1:-1]
        r = []
        for i in range(0, len(S) - 1):
            if self.isValid(S, 0, i) and self.isValid(S, i+1, len(S) - 1):
                left = self.gen(S, 0, i)
                right = self.gen(S, i+1, len(S) - 1)
                # print("left", left)
                # print("right", right)
                for s1 in left:
                    for s2 in right:
                        r.append("(%s, %s)" % (s1, s2))

        # print(r)
        return r


class TestSolution(unittest.TestCase):
    def test_trim(self):
        self.assertEqual(trim("  1234"), "1234")
        self.assertEqual(trim("1234  "), "1234")
        self.assertEqual(trim("   1234  "), "1234")
        self.assertEqual(trim("xxx1234xxx", "x"), "1234")

    def test_gen(self):
        s = Solution()
        self.assertListEqual(s.gen("12", 0, 1), ["12", "1.2"])
        self.assertListEqual(s.gen("100", 0, 2), ["100"])
        self.assertListEqual(s.gen("001", 0, 2), ["0.01"])
        self.assertListEqual(s.gen("0", 0, 0), ["0"])
        self.assertListEqual(s.gen("000001", 0, 5), ["0.00001"])

    def test_ambiguousCoordinates(self):
        s = Solution()
        l = s.ambiguousCoordinates("(123)")
        r = ["(1, 23)", "(12, 3)", "(1.2, 3)", "(1, 2.3)"]
        l.sort(), r.sort()
        self.assertListEqual(l, r)
        l = s.ambiguousCoordinates("(00011)")
        r = ["(0.001, 1)", "(0, 0.011)"]
        l.sort(), r.sort()
        self.assertListEqual(l, r)
        l = s.ambiguousCoordinates("(0123)")
        r = ["(0, 123)", "(0, 12.3)", "(0, 1.23)",
             "(0.1, 23)", "(0.1, 2.3)", "(0.12, 3)"]
        l.sort(), r.sort()
        self.assertListEqual(l, r)
        l = s.ambiguousCoordinates("(100)")
        r = ["(10, 0)"]
        l.sort(), r.sort()
        self.assertListEqual(l, r)
        l = s.ambiguousCoordinates("(0000001)")
        r = ["(0, 0.00001)"]
        l.sort(), r.sort()
        self.assertListEqual(l, r)


if __name__ == "__main__":
    unittest.main()
